### ISSUE TITLE
[stable/traefik] Allow setting targetPort for https

### ISSUE THREAD

**Issue by jangrewe** [helm/charts#885]
I'm trying to run Traefik on AWS with a SSL-terminating ELB, which works fine for plain HTTP, but unfortunately HTTPS gets forwarded the the HTTPS port of the Traefik pod, which is not in use.
I'd like to have the ELB do the SSL termination for HTTPS and then just forward the plain traffic to the HTTP port of Traefik.

I am able to do this manually by setting
```
spec:
  ports:
    name: https
      targetPort: 80
```
instead of `targetPort: 443` (the default), but manual changes are not ideal, of course.

I know that i could enable SSL in Traefik and have it do the SSL termination, but for various reasons i prefer to have the SSL termination done by the ELB.

**Comment by krancour** [comment]
SSL termination by the ELB is usually asking for trouble because each front-end listener can only support one certificate. Over time, this may lead you to create multiple ELBs and incur the expense thereof-- which negates the cost savings you typically realize by having a single point of ingress (the ingress controller) as the only service requiring an external load balancer.

I trust you have your reasons, so putting all of that aside...

What you're attempting to do can be done entirely through ELB configuration and doesn't require changes to the chart. You need to use `HTTPS` as the "Load Balancer Protocol" and `443` as the "Load Balancer Port", but when configuring the back-end that listener will talk to, use `HTTP` as the "Instance Protocol" and whatever node port is mapped to port 80 of the Traefik service as the "Instance Port".

**Comment by jangrewe** [comment]
Being able to use only one cert per ELB is not an issue, as we're using a wildcard cert that handles this, but i understand your concern.

I probably forgot to mention that the ELBs are automatically created by Kubernetes, not manually (`serviceType: LoadBalancer`), so i could of course manually change the ELBs after deployment, but that defeats the purpose of using this chart. ;-)

It could be easily accomplished by extending the template `stable/traefik/templates/service.yaml` with something like
```
{{- if .Values.ssl.targetPort }}
targetPort: {{ .Values.ssl.targetPort }}
{{- end }}
```
and then defining that value when installing the chart.

**Comment by krancour** [comment]
> I probably forgot to mention that the ELBs are automatically created by Kubernetes, not manually (serviceType: LoadBalancer), so i could of course manually change the ELBs after deployment, but that defeats the purpose of using this chart. ;-)

You have to edit your ELB manually anyway, because the ELB that k8s AWS cloud provider will create for you in response to Traefik's service being one of `type: LoadBalancer` will be a TCP-based listener (not an HTTP/S based one) and also isn't going to include your certificate.

**Comment by jangrewe** [comment]
Uhm, no, i don't? 

I'm using these nifty `service.beta.kubernetes.io/aws-load-balancer-*` annotations that take care of using the right cert for the ELB (and ssl port/ backend protocol).

**Comment by krancour** [comment]
> I'm using these nifty service.beta.kubernetes.io/aws-load-balancer-* annotations that take care of using the right cert for the ELB (and ssl port/ backend protocol).

Ah yes... I forgot about those... and I actually implemented the annotation for configuring idle timeout...

I'm still confused, because shouldn't those allow you to achieve the ELB configuration I had described?

**Comment by jangrewe** [comment]
Unfortunately not, at least as far as my understanding of the source goes, as they only allow you to set the backend protocol, and what (listener) port should be used for SSL on the ELB - but what port the SSL listeners should talk to on the backend is set dynamically by Kubernetes / the definition of the service, and as that's 443 in this chart, it uses the matching port for that on the node. Setting `targetPort: 80` allows you to override it, but that can't be done without changing the template.

**Comment by krancour** [comment]
Ah... I suppose that makes sense, because the mapping from a given listener port to a given container port is _already_ defined by the service. It wouldn't make too much sense to have an annotation for it as well.

Ok... so I understand the need now. Thank you for your patience in getting me there.

We should definitely consider this, but I do think it's a bit of an outlying use case. So before rushing into adding a knob to the chart that very few will ever turn, can you elaborate at all on the choice to terminate SSL at the ELB? (Since this feature request is a direct consequence of that.) This chart makes it easy to add your wildcard cert as Traefik's default certificate. It's probably easier than adding extra annotations to the service, in fact. If you _were_ able to take advantage of that, it would also have the added benefit of keeping traffic secure all the way to the cluster.

Bottom line is... with it being as easy as it is, there must be some reason for not doing it. I would just want to understand that before adding this feature.

**Comment by jangrewe** [comment]
The reason i was trying to have the ELB do the SSL termination was that i didn't want to manually export the cert we have in ACM (if that even possible) and put that into the config for Traefik, and i was under the impression that an ELB listener wouldn't work with a self-signed cert, like the one that this chart includes - so i would definitely need a valid cert for the ELB -> K8s communication.

I had SSL enabled in Traefik before, but it wasn't working - the ELB kept sending empty replies when i was trying to connect via HTTPS, while it worked fine via HTTP.
I gave it another shot, and also changed the service annotation `service.beta.kubernetes.io/aws-load-balancer-backend-protocol` to `https`, et voilá... seems like ELBs *do* work with self-signed certs!
And the ELB still does the SSL termination, because i get our wildcard cert on the client.

**Comment by krancour** [comment]
So you're terminating and re-initiation SSL. As long as you're ok with that, awesome!

**Comment by blackstar257** [comment]
I too just ran across this issue. It would be nice to have an if statement for the elb service. Made a PR with example

**Comment by blackstar257** [comment]
My pull request/merge fixes this for me atleast. It doesn't let you specify the exact port, but thats okay for my purposes.

**Comment by tkh** [comment]
I was looking for the same sort of behaviour and ended up doing my own deploy of Traefik to get it going. This is the one thing stopping me from running this chart. 

My approach at the moment is to have the following in the traefik.toml:
```
defaultEntryPoints = ["http","https"]

[entryPoints]
  [entryPoints.http]
  address = ":80"
     [entryPoints.http.redirect]
       regex = "^http://(.*)"
       replacement = "https://$1"
   [entryPoints.https]
   address = ":443"
```

and the following in the ELB over in AWS (driven by my Traefik service with AWS annotations and port config):
```
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:acm:XXXX"
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "http"

spec:
  type: LoadBalancer
  selector:
    app: traefik-proxy
    tier: proxy
  ports:
  - name: http
    port: 80
    targetPort: 80
  - name: https
    port: 443
    targetPort: 80
```

![screenshot from 2017-05-15 16-48-20](https://cloud.githubusercontent.com/assets/6788002/26066345/6e97faa8-398e-11e7-9d66-651078168ec6.png)

My end goal is to use the certs from AWS, let the ELB deal with all of that and ensure nobody arrives/remains on port 80 when accessing anything.

If there is a better way to achieve that (using blackstar's change?) I have not figured it out but I would love to hear it and use this chart to do it. Getting even this far has been a massive challenge which I had to solve manually.

**Comment by krancour** [comment]
@tkh the fix for this was merged in #1017. This issue should have been closed.

**Comment by lachie83** [comment]
closing as requested
